File: bench-java/src/main/java/bench/Runner.java
package bench;

import com.hazelcast.core.HazelcastInstance;
import org.HdrHistogram.ConcurrentHistogram;
import org.HdrHistogram.Histogram;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Semaphore;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Load-generation drivers — mirror of hazelrust-bench/src/runner.rs. Closed-loop
 * uses C platform threads each issuing → awaiting → re-issuing (so outstanding
 * ops == C, matching the Rust tokio model). Open-loop uses a fixed-rate clock
 * with C-permit backpressure and coordinated-omission-correct latency
 * (completion − scheduled-start).
 */
public final class Runner {

    static final long LO = 1, HI = 60_000_000_000L; // 1 ns .. 60 s
    static final int SIG = 3;

    public static final class Errors {
        public long total, timeouts, server, connection, other;
        synchronized void bump(int code) {
            total++;
            switch (code) { case 1: timeouts++; break; case 2: connection++; break; case 3: server++; break; default: other++; }
        }
        void add(Errors o) { total += o.total; timeouts += o.timeouts; server += o.server; connection += o.connection; other += o.other; }
    }

    public static final class RunResult {
        public Histogram hist;
        public long ops;
        public Errors errors = new Errors();
        public List<Double> series = new ArrayList<>();
        public boolean saturated;
        public double measureSecs;
    }

    static int classify(Throwable e) {
        String s = (e.toString() + " " + (e.getMessage() == null ? "" : e.getMessage())).toLowerCase();
        if (s.contains("timeout") || s.contains("timed out")) return 1;
        if (s.contains("connection") || s.contains("connect") || s.contains("closed") || s.contains("disconnect")) return 2;
        if (s.contains("exception") || s.contains("server") || s.contains("hazelcast")) return 3;
        return 0;
    }

    static void record(Histogram h, long ns) { h.recordValue(Math.min(ns, HI)); }

    public static RunResult runClosed(HazelcastInstance client, Manifest.Cell cell, long seed,
                                       long warmupNs, long measureNs) throws InterruptedException {
        int c = Math.max(1, cell.concurrency);
        AtomicLong counter = new AtomicLong();
        long start = System.nanoTime();
        long warmupEnd = start + warmupNs;
        long measureEnd = warmupEnd + measureNs;

        Histogram[] hists = new Histogram[c];
        long[] opsArr = new long[c];
        Errors[] errsArr = new Errors[c];
        Thread[] threads = new Thread[c];

        for (int w = 0; w < c; w++) {
            final int wid = w;
            threads[w] = new Thread(() -> {
                Ops.Planner planner = new Ops.Planner(cell, wid, seed);
                Ops.Handles h = Ops.Handles.build(client, cell, wid);
                Histogram hist = new Histogram(LO, HI, SIG);
                long ops = 0;
                Errors errs = new Errors();
                while (true) {
                    long now = System.nanoTime();
                    if (now >= measureEnd) break;
                    boolean measuring = now >= warmupEnd;
                    Ops.Plan p = planner.plan();
                    long t0 = System.nanoTime();
                    try {
                        h.exec(p);
                        counter.incrementAndGet();
                        if (measuring) { record(hist, System.nanoTime() - t0); ops++; }
                    } catch (Throwable e) {
                        if (measuring) errs.bump(classify(e));
                    }
                }
                hists[wid] = hist;
                opsArr[wid] = ops;
                errsArr[wid] = errs;
            }, "bench-w" + w);
            threads[w].start();
        }

        // monitor: per-second cumulative samples for the throughput series
        List<double[]> samples = new ArrayList<>();
        Thread mon = new Thread(() -> {
            while (true) {
                long now = System.nanoTime();
                double elapsed = (now - start) / 1e9;
                synchronized (samples) { samples.add(new double[]{elapsed, counter.get()}); }
                if (now >= measureEnd) break;
                try { Thread.sleep(1000); } catch (InterruptedException ignored) { break; }
            }
        }, "bench-mon");
        mon.start();

        for (Thread t : threads) t.join();
        mon.join();

        Histogram merged = new Histogram(LO, HI, SIG);
        long totalOps = 0;
        Errors errors = new Errors();
        for (int w = 0; w < c; w++) {
            if (hists[w] != null) merged.add(hists[w]);
            totalOps += opsArr[w];
            if (errsArr[w] != null) errors.add(errsArr[w]);
        }

        RunResult rr = new RunResult();
        rr.hist = merged;
        rr.ops = totalOps;
        rr.errors = errors;
        rr.measureSecs = measureNs / 1e9;
        rr.series = perSecondSeries(samples, warmupNs / 1e9, (warmupNs + measureNs) / 1e9);
        rr.saturated = false;
        return rr;
    }

    public static RunResult runOpen(HazelcastInstance client, Manifest.Cell cell, long seed,
                                    long warmupNs, long measureNs, double targetRate) throws InterruptedException {
        int c = Math.max(1, cell.concurrency);
        double intervalNs = 1.0e9 / targetRate;
        Semaphore sem = new Semaphore(c);
        ExecutorService pool = Executors.newFixedThreadPool(c);
        Ops.Handles[] lanes = new Ops.Handles[c];
        for (int w = 0; w < c; w++) lanes[w] = Ops.Handles.build(client, cell, w);
        Ops.Planner planner = new Ops.Planner(cell, 0, seed);

        ConcurrentHistogram hist = new ConcurrentHistogram(LO, HI, SIG);
        AtomicLong ops = new AtomicLong();
        Errors errors = new Errors();

        long start = System.nanoTime();
        long warmupNsAbs = warmupNs;
        long totalNs = warmupNs + measureNs;
        long k = 0;
        double maxLagNs = 0;

        while (true) {
            double schedOff = k * intervalNs;
            if (schedOff >= totalNs) break;
            long scheduled = start + (long) schedOff;
            long now = System.nanoTime();
            if (scheduled > now) {
                long sleep = scheduled - now;
                try { Thread.sleep(sleep / 1_000_000L, (int) (sleep % 1_000_000L)); } catch (InterruptedException ignored) {}
            } else {
                double lag = now - scheduled;
                if (lag > maxLagNs) maxLagNs = lag;
            }
            boolean measuring = schedOff >= warmupNsAbs;
            sem.acquire();
            final Ops.Handles h = lanes[(int) (k % c)];
            final Ops.Plan p = planner.plan();
            final long sched = scheduled;
            pool.submit(() -> {
                try {
                    h.exec(p);
                    long lat = System.nanoTime() - sched;
                    if (measuring) { record(hist, lat); ops.incrementAndGet(); }
                } catch (Throwable e) {
                    if (measuring) errors.bump(classify(e));
                } finally {
                    sem.release();
                }
            });
            k++;
        }
        pool.shutdown();
        // drain outstanding
        sem.acquire(c);

        RunResult rr = new RunResult();
        rr.hist = hist;
        rr.ops = ops.get();
        rr.errors = errors;
        rr.measureSecs = measureNs / 1e9;
        rr.saturated = maxLagNs > Math.max(intervalNs * 5.0, 50_000_000.0);
        return rr;
    }

    static List<Double> perSecondSeries(List<double[]> samples, double lo, double hi) {
        List<Double> out = new ArrayList<>();
        for (int i = 0; i + 1 < samples.size(); i++) {
            double t0 = samples.get(i)[0], c0 = samples.get(i)[1];
            double t1 = samples.get(i + 1)[0], c1 = samples.get(i + 1)[1];
            double mid = (t0 + t1) / 2.0;
            if (mid >= lo && mid <= hi && t1 > t0) out.add((c1 - c0) / (t1 - t0));
        }
        return out;
    }
}

File: hazelrust-bench/src/runner.rs
//! Load-generation drivers: closed-loop (max throughput / service time) and
//! open-loop (latency under a fixed arrival rate, coordinated-omission correct).

use anyhow::Result;
use hazelcast_client::HazelcastClient;
use hdrhistogram::Histogram;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;
use std::time::{Duration, Instant};
use tokio::sync::{mpsc, Semaphore};

use crate::manifest::Cell;
use crate::ops::{Handles, Planner};
use crate::record::{classify_code, Errors};

pub struct RunResult {
    pub hist: Histogram<u64>,
    pub ops: u64,
    pub errors: Errors,
    pub series: Vec<f64>,
    pub saturated: bool,
    pub measure_secs: f64,
}

fn new_hist() -> Histogram<u64> {
    // 1 ns .. 60 s, 3 significant figures (methodology §9.1).
    Histogram::<u64>::new_with_bounds(1, 60_000_000_000, 3).expect("hist bounds")
}

/// Closed-loop: C workers each issue → await → issue. Measures max throughput
/// and service time. Under-reports tail latency under saturation (CO) — use
/// open-loop for authoritative latency.
pub async fn run_closed(
    client: &Arc<HazelcastClient>,
    cell: &Cell,
    seed: u64,
    warmup: Duration,
    measure: Duration,
) -> Result<RunResult> {
    let c = cell.concurrency.max(1);
    let counter = Arc::new(AtomicU64::new(0));
    let start = Instant::now();
    let warmup_end = start + warmup;
    let measure_end = warmup_end + measure;

    // Monitor task: sample the global counter each second to build a per-second
    // throughput series for the measurement window (stability metric §9.2).
    let mon_counter = counter.clone();
    let monitor = tokio::spawn(async move {
        let mut samples: Vec<(f64, u64)> = Vec::new();
        loop {
            let now = Instant::now();
            if now >= measure_end {
                samples.push((
                    now.duration_since(start).as_secs_f64(),
                    mon_counter.load(Ordering::Relaxed),
                ));
                break;
            }
            samples.push((
                now.duration_since(start).as_secs_f64(),
                mon_counter.load(Ordering::Relaxed),
            ));
            tokio::time::sleep(Duration::from_secs(1)).await;
        }
        samples
    });

    let mut tasks = Vec::with_capacity(c as usize);
    for w in 0..c {
        let mut planner = Planner::new(cell, w, seed);
        let h = Handles::build(client, cell, w).await?;
        let counter = counter.clone();
        tasks.push(tokio::spawn(async move {
            let mut hist = new_hist();
            let mut ops = 0u64;
            let mut errs = Errors::default();
            loop {
                let now = Instant::now();
                if now >= measure_end {
                    break;
                }
                let measuring = now >= warmup_end;
                let plan = planner.plan();
                let t0 = Instant::now();
                match h.exec(plan).await {
                    Ok(()) => {
                        counter.fetch_add(1, Ordering::Relaxed);
                        if measuring {
                            let dt = t0.elapsed().as_nanos() as u64;
                            hist.saturating_record(dt);
                            ops += 1;
                        }
                    }
                    Err(e) => {
                        if measuring {
                            errs.bump(classify_code(&e));
                        }
                    }
                }
            }
            (hist, ops, errs)
        }));
    }

    let mut merged = new_hist();
    let mut total_ops = 0u64;
    let mut errors = Errors::default();
    for t in tasks {
        let (h, o, e) = t.await?;
        merged.add(h).ok();
        total_ops += o;
        errors += e;
    }
    let samples = monitor.await?;
    let series = per_second_series(
        &samples,
        warmup.as_secs_f64(),
        (warmup + measure).as_secs_f64(),
    );

    Ok(RunResult {
        hist: merged,
        ops: total_ops,
        errors,
        series,
        saturated: false,
        measure_secs: measure.as_secs_f64(),
    })
}

/// Busy-wait until `scheduled` with µs accuracy: sleep the bulk, spin the last
/// ~1 ms. Runs on a DEDICATED OS thread (not a tokio worker) so spinning never
/// starves the async runtime. This avoids the ~1 ms timer-granularity burst
/// artifact that a sleep-based pacer would inject into the CO-corrected latency.
fn precise_wait_until(scheduled: Instant) {
    loop {
        let now = Instant::now();
        if now >= scheduled {
            return;
        }
        let rem = scheduled - now;
        if rem > Duration::from_micros(1500) {
            std::thread::sleep(rem - Duration::from_micros(1000));
        } else {
            std::hint::spin_loop();
        }
    }
}

/// Open-loop: a fixed-rate clock issues requests regardless of completion (up
/// to C outstanding). Latency = completion − scheduled-start (coordinated
/// omission correct). A dedicated scheduler thread paces with µs accuracy and
/// dispatches onto the tokio runtime. Flags `saturated` if the generator falls
/// behind (had to wait for an outstanding-op permit, i.e. the rate exceeds
/// client capacity).
pub async fn run_open(
    client: &Arc<HazelcastClient>,
    cell: &Cell,
    seed: u64,
    warmup: Duration,
    measure: Duration,
    target_rate: f64,
) -> Result<RunResult> {
    let c = cell.concurrency.max(1) as usize;
    let interval_ns = 1.0e9 / target_rate;
    let sem = Arc::new(Semaphore::new(c));
    let mut lanes = Vec::with_capacity(c);
    for w in 0..c {
        lanes.push(Handles::build(client, cell, w as u32).await?);
    }
    let planner = Planner::new(cell, 0, seed);

    let warmup_ns = warmup.as_nanos() as f64;
    let total_ns = (warmup + measure).as_nanos() as f64;

    let (tx, mut rx) = mpsc::unbounded_channel::<(u64, bool, std::result::Result<(), u8>)>();
    let recorder = tokio::spawn(async move {
        let mut hist = new_hist();
        let mut ops = 0u64;
        let mut errs = Errors::default();
        while let Some((lat, measuring, res)) = rx.recv().await {
            match res {
                Ok(()) => {
                    if measuring {
                        hist.saturating_record(lat);
                        ops += 1;
                    }
                }
                Err(code) => {
                    if measuring {
                        errs.bump(code);
                    }
                }
            }
        }
        (hist, ops, errs)
    });

    let handle = tokio::runtime::Handle::current();
    let sched_tx = tx.clone();
    let scheduler = std::thread::spawn(move || {
        let mut planner = planner;
        let start = Instant::now();
        let mut k = 0u64;
        loop {
            let sched_off = k as f64 * interval_ns;
            if sched_off >= total_ns {
                break;
            }
            let scheduled = start + Duration::from_nanos(sched_off as u64);
            precise_wait_until(scheduled);
            let measuring = sched_off >= warmup_ns;
            // Acquire an outstanding-op permit; spin if none. A transient wait
            // (e.g. a GC pause filling the C window) is a LATENCY event captured
            // in the histogram — not saturation. Sustained inability to keep up
            // shows as wall-clock overrun (computed after the loop).
            let permit = loop {
                match sem.clone().try_acquire_owned() {
                    Ok(p) => break p,
                    Err(_) => std::hint::spin_loop(),
                }
            };
            let h = lanes[(k as usize) % c].clone();
            let plan = planner.plan();
            let txc = sched_tx.clone();
            handle.spawn(async move {
                let _permit = permit;
                let res = h.exec(plan).await;
                let lat = Instant::now()
                    .saturating_duration_since(scheduled)
                    .as_nanos() as u64;
                let code = match &res {
                    Ok(()) => Ok(()),
                    Err(e) => Err(classify_code(e)),
                };
                let _ = txc.send((lat, measuring, code));
            });
            k += 1;
        }
        // Overrun = how far past the intended schedule the generator ended up.
        // Bounded transient lag recovers (overrun ~0); only a generator that
        // genuinely cannot sustain the rate ends materially behind.
        Instant::now().saturating_duration_since(start).as_nanos() as f64 - total_ns
    });
    drop(tx); // scheduler holds the last sender clone; recorder ends when it + tasks finish
    let (hist, ops, errs) = recorder.await?;
    let overrun_ns = tokio::task::spawn_blocking(move || scheduler.join().unwrap()).await?;
    // Saturated only if the generator finished materially behind the schedule.
    let saturated = overrun_ns > (total_ns * 0.05).max(100_000_000.0);

    Ok(RunResult {
        hist,
        ops,
        errors: errs,
        series: Vec::new(),
        saturated,
        measure_secs: measure.as_secs_f64(),
    })
}

/// Convert (elapsed_s, cumulative_count) samples into per-second throughput
/// values that fall within the measurement window [warmup, warmup+measure].
fn per_second_series(samples: &[(f64, u64)], lo: f64, hi: f64) -> Vec<f64> {
    let mut out = Vec::new();
    for win in samples.windows(2) {
        let (t0, c0) = win[0];
        let (t1, c1) = win[1];
        let mid = (t0 + t1) / 2.0;
        if mid >= lo && mid <= hi && t1 > t0 {
            out.push((c1.saturating_sub(c0)) as f64 / (t1 - t0));
        }
    }
    out
}
